src/app/services/contracts.service.ts

```diff
@@ -0,0 +1,45 @@
+import {Injectable} from '@angular/core';
+import {HttpClient} from '@angular/common/http';
+import {Observable} from 'rxjs';
+import {environment} from '../../environments/environment';
+import {catchError, map} from 'rxjs/operators';
+import {SnackService} from './snack.service';
+import {Vault} from '../models/vault';
+import {Token} from '../models/token';
+import {Pool} from '../models/pool';
+import {Pair} from '../models/pair';
+
+/**
+ * Usage:
+ * const service = new ContractsService(httpClient, snackService);
+ * service.getContracts(Vault)  => Array<Vault>
+ * service.getContracts(Pool)  => Array<Pool>
+ * service.getContracts(Token)  => Array<Token>
+ *
+ */
+@Injectable({
+    providedIn: 'root'
+})
+export class ContractsService {
+
+    private urlPrefix = 'contracts';
+    private typePaths = new Map<new () => Vault|Pool|Token|Pair, string>(
+        [[Vault, 'vault'],
+            [Pool, 'pool'],
+            [Token, 'token'],
+            [Pair, 'unipair']]
+    );
+
+    constructor(private http: HttpClient, private snackService: SnackService) {
+
+    }
+
+    getContracts<T extends Vault|Pool|Token|Pair>(type: new () => T): Observable<T[]> {
```

Comment:
we can load it once and use cache 
implement lazy loading (when internal cache is empty - load new values, after that use only cached values)
and feel free implement the solution in all places
remove all static addresses and the vault name list

Reply:
👍